lib/install/vchlog/buffered_pipe.go

```diff
@@ -37,24 +41,28 @@ func NewBufferedPipe() *BufferedPipe {
 	var m sync.Mutex
 	c := sync.NewCond(&m)
 	return &BufferedPipe{
-		buffer: bytes.NewBuffer(nil),
-		c:      c,
-		closed: false,
+		buffer:      bytes.NewBuffer(nil),
+		c:           c,
+		readClosed:  false,
+		writeClosed: false,
+		readerReady: false,
 	}
 }
 
 // Read is blocked until a writer in the queue is done writing (until data is available)
 func (bp *BufferedPipe) Read(data []byte) (n int, err error) {
 	bp.c.L.Lock()
 	defer bp.c.L.Unlock()
+	defer bp.c.Broadcast()
 
-	// pipe closed, drop all left-over data
-	if bp.closed {
-		return 0, io.EOF
-	}
-	for bp.buffer.Len() == 0 && !bp.closed {
+	bp.readerReady = true // now we have a valid consumer to switch to flushing in Close
```

Comment:
What happens if we had a reader at one point, but no longer have one?

This shouldn't happen with the way this code is currently used, but in theory I think something like this could happen:
```
bp = NewBufferedPipe()
bp.Write([]byte("Hello World"))
b = make([]byte, 5)
bp.Read(b) // Read "Hello", caused readerReady = true
bp.Close() // Waits forever for someone to read " World"
```

Maybe this isn't a case we can reasonably handle? (Or maybe we handle this with some sort of timeout on the flush?

Reply:
Yes, it would probably be a case we can't reasonably handle. The read is from the http connection established by govc with the datastore, so the early abortion of read only happens if the datastore connection is cut down when consuming the logs. In this case, `datastore.Upload` will return an error. However I'm not sure how to treat this connection failure reasonably, it could be the problem with VCH, or the problem from client side.
A possible way to handle this is that, if upload returns an error in `vchlogger.go`, we send a signal to the underlying pipe telling it to drop any leftover messages.

Reply:
In general usage, it's hard to tell if someone is still calling read() continuously. If someone calls read() at one point and stop, the pipe won't know about this (pipe isn't keeping track of the status of caller). Adding a timeout would be a possible workaround

Reply:
Ironically adding a `WriteTo` method would probably allow us to resolve this, but in a manner that's not consistent with the pure Reader behaviour.
WriteTo shouldn't return at all until the copy has completely finished, and you would get an error from the writer that can be used to purge any remaining data.